### PR TITLE
**Fix:** 1078 code examples

### PR DESCRIPTION
**Pull request by bradd123**
# Summary

# Related issue

# To be tested

Me
- [x] No error or warning in the console on `localhost:6060`

Tester 1

- [ ] Things look good on the demo.
  

Tester 2

- [ ] Things look good on the demo.

### CODE REVIEW

File: src/Autocomplete/README.md [hunk 1]
```diff
@@ -43,6 +43,11 @@ const MyComponent = () => {
       })
   }, [dogToFetch])
 
+  const onResultClickUtil = result => {
```
Comment (TejasQ) [inline src/Autocomplete/README.md]: Use `useCallback` here to not recompute the function every time.

File: src/CardSection/CardSection.tsx [hunk 2]
```diff
@@ -9,7 +9,7 @@ import { headerHeight } from "../utils/constants"
 
 export interface CardSectionProps extends DefaultProps, DragProps {
   /** Column title */
-  title?: string
+  title?: React.ReactElement<any> | string
```
Comment (TejasQ) [inline src/CardSection/CardSection.tsx]: Would `React.ReactNode` work here?

File: src/ContextMenu/README.md [hunk 3]
```diff
@@ -62,7 +62,7 @@ import * as React from "react"
 import { AddIcon, ContextMenu, ContextMenuProps, Contact, Hint, Button } from "@operational/components"
 
 /* Anything can be a label now, even some DIV you style yourself */
-const MyLabelContainer = ({ children, style }) => (
+const MyLabelContainer = ({ children, style = {} }) => (
```
Comment (TejasQ) [inline src/ContextMenu/README.md]: We can just remove the `style` prop and its spread on the next line.

File: src/Page/README.md [hunk 4]
```diff
@@ -200,7 +200,7 @@ const Tab = props => (
   </PageContent>
 )
 
-class Router extends React.Component {
+class Router extends React.Component<{}, any> {
```
Comment (TejasQ) [inline src/Page/README.md]: Would you be willing to rewrite this with hooks? It could sharpen your hooks skills too!

File: src/Select/README.md [hunk 5]
```diff
@@ -22,7 +22,7 @@ const options = [
 ]
 
 const MyComponent = () => {
-  const [value, setValue] = React.useState("one")
+  const [value, setValue] = React.useState < string > "one"
```
Comment (TejasQ) [inline src/Select/README.md]: Is this valid TS? It looks more like elm...

Comment (bradd123) [reply src/Select/README.md]: That was prettier in work. I wanted to write `React.useState<string>("one")`.

File: src/Page/README.md (hunk 4 above)
Comment (bradd123) [reply src/Page/README.md]: Cool. will do that.

**Comment by TejasQ** [comment]
@bradd123 what up do you need help with anythang

**Comment by TejasQ** [comment]
A week later, paging captiain @bradd123. How's it going? Can I help with anything?

**Comment by bradd123** [comment]
Hi, Most of the things are done. I am having difficulty understanding the following component type errors. `DataTable`, `DataTableInput`, `HeaderMenu`, `LabelText`, `Page`, `Select`, `Tree`, `FinePrint`.

**Comment by bradd123** [comment]
I will create a separate PR to convert `import * as React from 'react'` to `import React from 'react'`.

**Comment by bradd123** [comment]
I have seen multiple type errors like this. `Type 'number' is not assignable to type 'SetStateAction<string>'` What is the solution for these types? Any good references would be welcome.

**Comment by TejasQ** [comment]
> I have seen multiple type errors like this. Type 'number' is not assignable to type 'SetStateAction<string>'

Pointing me to where you see these errors would be a start to helping me help you. 💞

**Comment by TejasQ** [comment]
To fix the type errors:

## About `DataTable`
- `DataTableFooter` apparently has no `onClick` prop! Odd! Maybe we should add one. Open an issue if you don't want to fix it in this PR.
- You'll need to rewrite the `getNewRows` function like so:

```diff
const getNewRows = (lastIndex = 0): Array<[string, number, number]> =>
  Array.from({ length: lastIndex > 1000000 ? 1 : (lastIndex || 1) * 10 }, (_, i) => [
    `Cell ${i + lastIndex + 1}`,
    Math.random(),
    Math.random(),
-  ])
+  ]) as Array<[string, number, number]>
```

TypeScript doesn't really infer types of matrices well enough so we need to add a type assertion here and everything will be fine.

## About `DataTableInput`
Its [`value` prop](https://github.com/contiamo/operational-ui/blob/7f3a10e1e24ff04b90a54940644d3c9340ebaa1c/src/Input/Input.tsx#L20) is a `string`, but in this README example, we're [passing it a `number`](https://github.com/contiamo/operational-ui/blame/7f3a10e1e24ff04b90a54940644d3c9340ebaa1c/src/DataTableInput/README.md#L11). The correct solution would be to coerce the `number` we pass to `useState` to a `string`:

```diff
-  const [age, setAge] = React.useState(new Date().getFullYear() - 1993)
+  const [age, setAge] = React.useState(String(new Date().getFullYear() - 1993))
```

## About `HeaderMenu`

TypeScript has no idea what [`projectOptions`](https://github.com/contiamo/operational-ui/blame/7f3a10e1e24ff04b90a54940644d3c9340ebaa1c/src/HeaderMenu/README.md#L9) is and it can't figure it out on its own. We need to help it out here by telling it:

```diff
- const projectOptions = [
+ const projectOptions: ContextMenuProps["items"] = [
```

## About `LabelText`

There is no valid `htmlFor` prop on `span` elements – [which it is](https://github.com/contiamo/operational-ui/blame/7f3a10e1e24ff04b90a54940644d3c9340ebaa1c/src/LabelText/LabelText.tsx#L4). The solution in this case is to remove the prop from the README.

## About `Page`
1. `styled` is not imported from Operational UI so TypeScript doesn't know what it is. The solution is to add it to the list of imports.
1. You can remove the whole `Actions` component and its import to fix the other error. It's an internal component that we should either externalize or unexpose.

## About `Select`

A `value` given to the first argument of `onChange` can be `String | Number | ...`. What you'll need to do is assert what the new value actually is when you call `setState` like this:

```diff
<Select
        id="basic"
        label="Basic"
        data-cy="basic-select"
        value={value}
        options={options}
        onChange={(newValue, lastChanged) => {
-          setValue(newValue)
+          setValue(newValue as string)
          setLastChanged(lastChanged)
        }}
      />
```

## About `Tree`

This is a little more involved. Let's make an issue and deal with it later.

## About `FinePrint`

This is an odd error. I opened an issue about it on emotion (https://github.com/emotion-js/emotion/issues/1431). Let's see what they say. We can open an issue for this and continue with the PR.

**Comment by bradd123** [comment]
Thanks for the detailed explanation. I will make the changes and push tonight.

**Comment by bradd123** [comment]
`HeaderMenu` has a weird problem. Directly editing works, but changing the file is not working. The following error is showing.

> Error: import or require() statements can be added only by editing a Markdown example file: @operational/components

**Comment by TejasQ** [comment]
I know this problem. I think @stereobooster once solved it really well. I wonder where his code is... 👀

**Comment by TejasQ** [comment]
Here!

https://github.com/contiamo/operational-ui/commit/728f18ffce5c595dd073308faeb38b58bc879de1

Apparently removing a type annotation resolves it.

**Comment by bradd123** [comment]
What if we need the type annotation? Remove annotation throws type errors.

**Comment by TejasQ** [comment]
I honestly don't know 😂

File: src/Card/Card.tsx [hunk 6]
```diff
@@ -13,6 +13,7 @@ export interface BaseProps extends DefaultProps {
    */
   fullSize?: boolean
   onClick?: (event: React.MouseEvent<HTMLDivElement>) => void
+  style?: React.CSSProperties
```
Comment (TejasQ) [inline src/Card/Card.tsx]: Wy do we need this?

**Comment by bradd123** [comment]
What is your preferred way to merge the code? Do you rebase or directly merging?

File: src/OperationalUI/README.md [hunk 7]
```diff
@@ -23,9 +23,9 @@ To set up routing that is automatically wired up to the `to` props of all nested
 
 ```jsx
 import * as React from "react"
-import { Button } from "@operational/components"
+import { Button, OperationalUI, OperationalUIProps } from "@operational/components"
 
-class RoutingComponent extends React.Component {
+class RoutingComponent extends React.Component<OperationalUIProps, any> {
```
Comment (TejasQ) [inline src/OperationalUI/README.md]: no `any`. Write a definition of the state. Should just be `{ path: string }`

Comment (TejasQ) [reply src/OperationalUI/README.md]: Even better, this could be rewritten as hooks if you want the knowledge/fun.

**Comment by bradd123** [comment]
The remaining components are `DataTable`, `HeaderMenu`, `LabelText`, `Tree`, and `FinePrint`. Each one has separate issues. We will deal with them separately. Please review the existing ones.

Next, I will create a separate PR that will convert all `import * as React from "react"` to `import React from "react"`. I have already changed compiler options in this PR.

File: src/Card/Card.tsx (hunk 6 above)
Comment (bradd123) [reply src/Card/Card.tsx]: We don't actually need that. I added that a week back. I will update this file and doc.

File: src/OperationalUI/README.md (hunk 7 above)
Comment (bradd123) [reply src/OperationalUI/README.md]: Sure, will convert it into a hook

Comment (bradd123) [reply src/OperationalUI/README.md]: Done

File: src/Card/Card.tsx (hunk 6 above)
Comment (bradd123) [reply src/Card/Card.tsx]: Updated this component.

**Comment by TejasQ** [comment]
Great! If you want to deal with all the other issues separately, would you go ahead and make issues for them so we can keep track of them and open corresponding PRs? That way we can merge this one.

**Comment by bradd123** [comment]
Sure, I will create separate issues for them.

**Comment by TejasQ** [comment]
Finally, it would be great if you filled this out:

image

Attention to detail is something we value and encourage on this project. I think once that's done we can take a train to merge city. 🚋

**Comment by bradd123** [comment]
I am having the following warnings in the console.
Operational_UI_and_Slack_-_RemoteIndian
Operational_UI
Operational_UI
Operational_UI

**Comment by TejasQ** [comment]
Right – those existed before. Great. I think then we're good.

**Pull request merged**